perf/intkey_workload/src/intkey_iterator.rs

```diff
@@ -0,0 +1,221 @@
+    incs_and_decs_per_set: u32,
+    set_value: u32,
+}
+
+impl IntKeyIterator {
+    pub fn new(num_names: usize, invalid: f32, seed: &[usize]) -> IntKeyIterator {
+        IntKeyIterator {
+            num_names: num_names,
+            invalid: invalid,
+
+            rng: SeedableRng::from_seed(seed),
+
+            names: HashMap::new(),
+            incs_and_decs_per_set: INCS_AND_DECS_PER_SET as u32,
+            set_value: HALF_MAX_VALUE,
+        }
+    }
+
+    fn gen_name(&mut self) -> String {
+        self.rng.gen_ascii_chars().take(MAX_NAME_LEN).collect()
+    }
+
+    /// Generate a RNG chosen inc or dec payload with the name supplied if not None,
+    /// else if name is None, choose a random name.
+    fn gen_inc_or_dec(&mut self, name: Option<String>) -> IntKeyPayload {
+        let verb;
+        match self.rng.gen::<bool>() {
+            true => verb = "inc",
+            false => verb = "dec",
+        }
+
+        let n;
+        match name {
+            Some(name) => n = name,
+            None => n = self.gen_name(),
+        }
+
+        IntKeyPayload {
+            name: n,
+            verb: verb.to_string(),
+            value: self.rng.gen_range(1, 10),
+        }
+    }
+}
+
+impl Iterator for IntKeyIterator {
+    type Item = IntKeyPayload;
+
+    /// First use RNG to choose between invalid and valid txn.
+    /// If valid, determine if there have been enough sets. If so choose between inc or dec.
+    /// If not generate a random name to set.
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.rng.gen_range(0.0, 1.0) >= self.invalid {
+            // Valid transaction
+            match self.names.len() < self.num_names {
+                true => {
+                    let name = self.gen_name();
+                    self.names.insert(name.clone(), 0);
+                    return Some(IntKeyPayload {
+                        name: name,
+                        verb: "set".to_string(),
+                        value: self.set_value,
+                    });
+                }
+                false => {
+                    let names_map_clone = self.names.clone();
+                    let index = self.rng.gen_range(0, self.names.len());
+                    let name = names_map_clone.keys().nth(index);
+                    let n;
+                    match name {
+                        Some(name) => {
+                            n = name.to_string();
+                            let value = self.names.get(name).unwrap_or(&0).to_owned();
+
+                            if value < self.incs_and_decs_per_set {
+                                self.names.insert(name.to_string(), value + 1);
+                            } else {
+                                self.names.remove(name);
+                            }
+                        }
+                        None => {
+                            warn!("There is a severe indexing bug in the intkey workload.");
```

Comment:
If this is an unrecoverable error, then it should probably just panic()

Reply:
Yeah, I agree. I will use `.expect` and write some more tests.